### PR TITLE
Update patching playbook to utilize kubernetes.core collection

### PR DESCRIPTION
**Pull request by PrymalInstynct**
Recommend replacing shell and command modules with kubernetes.core.k8s_drain

- The kubernetes.core.k8s_drain module is well supported and conducts the checks being done as a part of the command and shell tasks already written in this playbook.

- In the upcoming kubernetes.core v2.5.0 collection release pod_selectors and label_selectors will be supported to make the drain process faster/more accurate **see commented example on lines 43-45.** [add ability to filter the list of pods to be drained by a pod label selector](https://github.com/ansible-collections/kubernetes.core/pull/606)

- Added Check to determine if the node requires a reboot based on installed packages before executing the reboot task
**NOTE: this is only applicable to Debian based nodes**

### CODE REVIEW

**Comment by onedr0p** [comment]
This looks awesome @PrymalInstynct were you able to test this and all seemed fine?

**Comment by PrymalInstynct** [comment]
So I did not test the conversion from template to playbooks, but I wrote [my own role](https://github.com/PrymalInstynct/k8s-inf-update) about a month ago that did the same thing along with sending notifications to discord based on the status of the various tasks so I could patch weekly on a cron and know when the cluster rebooted. The logic is sound in my role and I copy pasted most of it, but I am happy to test this PR later today/tomorrow.

File: bootstrap/templates/ansible/playbooks/cluster-rollout-update.yaml.j2 [hunk 1]
```diff
@@ -25,46 +25,45 @@
...
 
-        - name: Wait for node to cordon
-          ansible.builtin.command: "kubectl get node {% raw %}{{ inventory_hostname }}{% endraw %} -o json"
-          register: wait_for_cordon
-          retries: 10
-          delay: 10
+        - name: "Drain {% raw %}{{ inventory_hostname }}{% endraw %}"
+          kubernetes.core.k8s_drain:
+            state: drain
+            name: {% raw %}{{ inventory_hostname }}{% endraw %}
```
Comment (onedr0p) [inline bootstrap/templates/ansible/playbooks/cluster-rollout-update.yaml.j2:37]:
```suggestion
            name: "{% raw %}{{ inventory_hostname }}{% endraw %}"
```

File: bootstrap/templates/ansible/playbooks/cluster-rollout-update.yaml.j2 [hunk 2]
```diff
@@ -25,46 +25,45 @@
...
-          retries: 10
-          delay: 10
-          delegate_to: "{% raw %}{{ groups['master'][0] }}{% endraw %}"
-          changed_when: false
-          until: not (kubectl_get_node['stdout'] | from_json).spec.unschedulable is defined
+        - name: "Uncordon {% raw %}{{ inventory_hostname }}{% endraw %}"
+          kubernetes.core.k8s_drain:
+            state: uncordon
+            name: "{% raw %}{{ inventory_hostname }}{% endraw %}"
+          delegate_to: localhost
```
Comment (onedr0p) [inline bootstrap/templates/ansible/playbooks/cluster-rollout-update.yaml.j2:69]: Should this be delegated to `{{ groups['master'][0] }}` like the others?

File: bootstrap/templates/ansible/playbooks/cluster-rollout-update.yaml.j2 [hunk 3]
```diff
@@ -25,46 +25,45 @@
         # If spec.unschedulable is defined then the node is cordoned
         - not (kubectl_get_node['stdout'] | from_json).spec.unschedulable is defined
       block:
-        - name: Cordon node
-          ansible.builtin.command: "kubectl cordon {% raw %}{{ inventory_hostname }}{% endraw %}"
+        - name: "Cordon {% raw %}{{ inventory_hostname }}{% endraw %}"
+          kubernetes.core.k8s_drain:
+            state: cordon
+            name: "{% raw %}{{ inventory_hostname }}{% endraw %}"
```
Comment (onedr0p) [inline bootstrap/templates/ansible/playbooks/cluster-rollout-update.yaml.j2:31]:
```suggestion
        - name: "Cordon {% raw %}{{ inventory_hostname }}{% endraw %}"
          kubernetes.core.k8s_drain:
            name: "{% raw %}{{ inventory_hostname }}{% endraw %}"
            kubeconfig: /etc/rancher/k3s/k3s.yaml
            state: cordon
```

File: bootstrap/templates/ansible/playbooks/cluster-rollout-update.yaml.j2 [hunk 4]
```diff
@@ -25,46 +25,45 @@
...
-        - name: Wait for node to cordon
-          ansible.builtin.command: "kubectl get node {% raw %}{{ inventory_hostname }}{% endraw %} -o json"
-          register: wait_for_cordon
-          retries: 10
-          delay: 10
+        - name: "Drain {% raw %}{{ inventory_hostname }}{% endraw %}"
+          kubernetes.core.k8s_drain:
+            state: drain
+            name: "{% raw %}{{ inventory_hostname }}{% endraw %}"
+            delete_options:
```
Comment (onedr0p) [inline bootstrap/templates/ansible/playbooks/cluster-rollout-update.yaml.j2:38]:
```suggestion
          kubernetes.core.k8s_drain:
            name: "{% raw %}{{ inventory_hostname }}{% endraw %}"
            kubeconfig: /etc/rancher/k3s/k3s.yaml
            state: drain
            delete_options:
```

File: bootstrap/templates/ansible/playbooks/cluster-rollout-update.yaml.j2 (hunk 3 above)
Comment (onedr0p) [reply bootstrap/templates/ansible/playbooks/cluster-rollout-update.yaml.j2:31]: Re-arranged the vars and included the kubeconfig location

File: bootstrap/templates/ansible/playbooks/cluster-rollout-update.yaml.j2 (hunk 4 above)
Comment (onedr0p) [reply bootstrap/templates/ansible/playbooks/cluster-rollout-update.yaml.j2:38]: Re-arranged the vars and included the kubeconfig location

File: bootstrap/templates/ansible/playbooks/cluster-rollout-update.yaml.j2 [hunk 5]
```diff
@@ -25,46 +25,45 @@
...
-          register: wait_for_uncordon
-          retries: 10
-          delay: 10
-          delegate_to: "{% raw %}{{ groups['master'][0] }}{% endraw %}"
-          changed_when: false
-          until: not (kubectl_get_node['stdout'] | from_json).spec.unschedulable is defined
+        - name: "Uncordon {% raw %}{{ inventory_hostname }}{% endraw %}"
+          kubernetes.core.k8s_drain:
+            state: uncordon
+            name: "{% raw %}{{ inventory_hostname }}{% endraw %}"
```
[the remark this answers is not in the document]
Comment (onedr0p) [reply bootstrap/templates/ansible/playbooks/cluster-rollout-update.yaml.j2:68]: Re-arranged the vars and included the kubeconfig location

**Review by onedr0p — CHANGES_REQUESTED** [review]

File: bootstrap/templates/ansible/playbooks/cluster-rollout-update.yaml.j2 (hunk 2 above)
Comment (PrymalInstynct) [reply bootstrap/templates/ansible/playbooks/cluster-rollout-update.yaml.j2]: Yep, committed fix for this. I have been using my dev box (localhost) to run my role on cron, didn't even think to use one of the control plane nodes.

**Comment by PrymalInstynct** [comment]
So you will notice I am still running the shell module in my role for the drain task because I can include the pod_selectors flag to ignore the app `rook-ceph-osd` which causes me the same problem you had above (similar issues exist with other object storage like Longhorn too). As mentioned in the initial post once kubernetes.core v2.5.0 is released this can be addressed in the k8s_drain module per my commented out example.

So we could keep using the command module for that task until v2.5.0 is published. I will let you make that call.

**Review by PrymalInstynct — COMMENTED** [review]
Makes sense, when I pulled the kubeconfig parameter out I assumed that a control plane node would automatically know where the kubeconfig file lived.

File: bootstrap/templates/ansible/playbooks/cluster-rollout-update.yaml.j2 [hunk 6]
```diff
@@ -25,46 +25,48 @@
...
+        - name: "Drain {% raw %}{{ inventory_hostname }}{% endraw %}"
+          kubernetes.core.k8s_drain:
+            name: "{% raw %}{{ inventory_hostname }}{% endraw %}"
+            kubeconfig: /etc/rancher/k3s/k3s.yaml
+            state: drain
+            delete_options:
+              delete_emptydir_data: true
+              ignore_daemonsets: true
+              terminate_grace_period: 300
+              wait_timeout: 600
```
Comment (onedr0p) [inline bootstrap/templates/ansible/playbooks/cluster-rollout-update.yaml.j2:44]:
```suggestion
              terminate_grace_period: 600
              wait_timeout: 900
              force: true
```

I think this should be good, some pods (loki) can take up-to 10 minutes to be killed.

**Review by onedr0p — CHANGES_REQUESTED** [review]

**Comment by PrymalInstynct** [comment]
## Original Playbook
I ran the original playbook and it has been sitting at the drain task on  the first node for 20 minutes.

I am using the following command to monitor what is going on with the pods
`kubectl get pods -A | grep -v Running `

And here is my output at the moment
```
NAMESPACE        NAME                                                           READY   STATUS      RESTARTS	 AGE
kyverno          kyverno-cleanup-admission-reports-28162750-7xbbq               0/1     Completed   0             7m14s
kyverno          kyverno-cleanup-cluster-admission-reports-28162750-q9ntj	0/1     Completed   0             7m14s
rook-ceph        rook-ceph-crashcollector-k8s-control-prod-0-6cc4b59d4c-c9kbf   0/1     Pending     0             20m
rook-ceph        rook-ceph-crashcollector-k8s-control-prod-0-79d4dc5fdf-fj5vd   0/1     Pending     0             20m
rook-ceph        rook-ceph-mon-a-554c94d5bd-5xn4p                               0/2     Pending     0             20m
rook-ceph        rook-ceph-osd-prepare-k8s-control-prod-1-zwwlc                 0/1     Completed   0             28m
rook-ceph        rook-ceph-osd-prepare-k8s-control-prod-2-74mjl                 0/1     Completed   0             28m
rook-ceph        rook-ceph-osd-prepare-k8s-worker-prod-3-jz78t                  0/1     Completed   0             28m
rook-ceph        rook-ceph-osd-prepare-k8s-worker-prod-4-5tswj                  0/1     Completed   0             28m
```

So I updated the original playbook to include the `--pod-selector='app!=rook-ceph-osd'` option and re-ran it.
This addressed the very long hang I was experiencing from rook-ceph but I ran into a hang when grafana was being migrated so I added `--grace-period=120` 
### The original playbook ran in  `9 minutes 3 seconds`

## Updated Playbook
Knowing that the `pod_selectors` parameter for the k8s_drain ansible module is not releases yet I used the same drain task as the original playbook, just to make it an apples vs apples comparison.
### The updated playbook ran in `9 minutes 20 seconds`

## Conclusion
I believe that the updated playbook is worth merging as it provides a very similar level of performance which using Ansible modules that meets their project standards of status/error checking, and idempotence.

I have committed an updated playbook template to this PR, that utilizes the command module for the drain task including the above modifications, the drain grace_period set to `300`, and has a complete k8s_drain task commented out ready for transition once the kubernetes.core v2.5.0 collection is released.

**Comment by PrymalInstynct** [comment]
No worrires, I think its ready to go now.

File: bootstrap/templates/ansible/playbooks/cluster-rollout-update.yaml.j2 [hunk 7]
```diff
@@ -25,24 +25,38 @@
         # If spec.unschedulable is defined then the node is cordoned
         - not (kubectl_get_node['stdout'] | from_json).spec.unschedulable is defined
       block:
-        - name: Cordon
-          ansible.builtin.command: "kubectl cordon {% raw %}{{ inventory_hostname }}{% endraw %}"
+        - name: "Cordon {% raw %}{{ inventory_hostname }}{% endraw %}"
```
Comment (onedr0p) [inline bootstrap/templates/ansible/playbooks/cluster-rollout-update.yaml.j2:28]:
```suggestion
        - name: Cordon
```

The node name is already included in the ansible output, we don't need it here or the other task names IMO.

File: bootstrap/templates/ansible/playbooks/cluster-rollout-update.yaml.j2 [hunk 8]
```diff
@@ -26,23 +26,37 @@
...
-          register: wait_for_cordon
-          retries: 10
-          delay: 10
-          delegate_to: "{% raw %}{{ groups['master'][0] }}{% endraw %}"
-          changed_when: false
-          until: (wait_for_cordon['stdout'] | from_json).spec.unschedulable
-
-        - name: Drain
-          ansible.builtin.command: "kubectl drain --ignore-daemonsets --delete-emptydir-data --force {% raw %}{{ inventory_hostname }}{% endraw %}"
+        - name: "Drain {% raw %}{{ inventory_hostname }}{% endraw %}"
```
Comment (onedr0p) [inline bootstrap/templates/ansible/playbooks/cluster-rollout-update.yaml.j2:35]:
```suggestion
        - name: Drain
```

File: bootstrap/templates/ansible/playbooks/cluster-rollout-update.yaml.j2 [hunk 9]
```diff
@@ -26,23 +26,37 @@
...
+          ansible.builtin.command: "kubectl drain --pod-selector='app!=rook-ceph-osd,app!=csi-attacher,app!=csi-provisioner' --ignore-daemonsets --delete-emptydir-data --force --grace-period=300 {% raw %}{{ inventory_hostname }}{% endraw %}"
           delegate_to: "{% raw %}{{ groups['master'][0] }}{% endraw %}"
           changed_when: false
+    
+        ### pod_selectors feature in upcoming kubernetes.core 2.5.0 ###
+        # - name: "Drain {% raw %}{{ inventory_hostname }}{% endraw %}"
```
Comment (onedr0p) [inline bootstrap/templates/ansible/playbooks/cluster-rollout-update.yaml.j2:41]:
```suggestion
        # - name: Drain
```

File: bootstrap/templates/ansible/playbooks/cluster-rollout-update.yaml.j2 [hunk 10]
```diff
@@ -60,16 +74,11 @@
...
-        - name: Uncordon
-          ansible.builtin.command: "kubectl uncordon {% raw %}{{ inventory_hostname }}{% endraw %}"
-          delegate_to: "{% raw %}{{ groups['master'][0] }}{% endraw %}"
-          changed_when: false
-
-        - name: Wait for uncordon
-          ansible.builtin.command: "kubectl get node {% raw %}{{ inventory_hostname }}{% endraw %} -o json"
-          retries: 10
-          delay: 10
+        - name: "Uncordon {% raw %}{{ inventory_hostname }}{% endraw %}"
```
Comment (onedr0p) [inline bootstrap/templates/ansible/playbooks/cluster-rollout-update.yaml.j2:79]:
```suggestion
        - name: Uncordon
```

**Review by onedr0p — CHANGES_REQUESTED** [review]

**Review by onedr0p — APPROVED** [review]

**Pull request merged**